### DISCUSSION TITLE
programatically select rows

### DISCUSSION THREAD

**Discussion by dnalor** [Evertras/bubble-table#137] [Q&A] [answered]
First of all, thanks for releasing this - very useful.

I'd like to programatically
 - set the highlighted row
 - selected / unselected rows
without rebuilding the table.

I tried:
```
for _, r := range m.archTable.SelectedRows() {
	r.Selected( false)
}
```
which doesn't work (presumably b/c the method returns values and not references?)

Am I missing something, or is this not possible?

**Comment by Evertras** [comment, marked as answer]
The rows returned by `SelectedRows` are just copies, so changing anything will not reflect back into the table.  It's a little misleading now that I'm double checking the docs.  Let me play with creating an example and potentially making some updates as well to make this easier/clearer.

**Reply by Evertras** [reply]
So internally this is actually doing a full copy of the rows, because with Bubble Tea the flow generally depends on these models being immutable.

https://github.com/Evertras/bubble-table/blob/main/table/update.go#L28

The `Selected()` method actually returns a new row with the selection applied to it.  So the recommended approach without any modifications would unfortunately be to rebuild the table, and have some row generation function that handles which ones are selected.  However, this feels very suboptimal.

Do you want to simply clear existing selections?  Or do you want some other way of selecting/deselecting specific rows based on an index?  I think we can add some helper functions here to make this easier.  Let's create an issue for this one.

**Reply by Evertras** [reply]
Regardless I do want to update the docs for `Selected` because the help text is misleading.

**Reply by Evertras** [reply]
https://github.com/Evertras/bubble-table/issues/139

https://github.com/Evertras/bubble-table/issues/140

**Reply by dnalor** [reply]
Thanks for the reply.
I suspected as much, but I haven't found public methods to set certain options for an _existing_ table (as opposed to doing so at creation time). 
Examples:
WithCurrentPage
Selected

I think methods like e.g.
SetCurrentPage
SetSelected
SetHighlightedRowIndex
would be quite useful, as it seems that currently one would have to rebuild the whole table every time one wants to change those options.

(I've seen 5ce8e35 , which certainly helps, but would still mean rebuilding the table)

**Reply by dnalor** [reply]
(sorry, should have read your newest comment first before commenting, which addresses what I've written)

**Reply by Evertras** [reply]
To clarify an important point, modifying an existing table is forbidden as Bubble Tea relies on all models being immutable.  Everything must be copied when a modification is made, otherwise things get weird for Bubble Tea if its assumptions are violated.  We can make that copy relatively efficient, but having exported functions that modify the underlying table's data is a no-go.

**Reply by dnalor** [reply]
Interesting, I wasn't aware of that.
Then it may not be worth your effort just for some corner cases.

**Reply by Evertras** [reply]
Gotcha... if there's something that's really a pain to do I'm always happy to figure out if there's some nice helper function we could add, like `WithAllRowsDeselected`.  Feel free to open an issue if you have any ideas for more. 🤘

**Reply by timcolson** [reply]
After starting with a simple table, I came here to ask how to update data for a selected row...

I found this Q&A and read the following: 
> _...modifying an existing table is forbidden as Bubble Tea relies on all models being **immutable**_

Ahhhh. Thx @Evertras! That helped Elm click for me! 

I've been struggling with changes not showing up in a List of items. 
Apparently, the LLM assistant I'm using is similarly confused about immutability. lol

TL;DR - To update one row, rebuild the model data for all the rows?

**Reply by Evertras** [reply]
Basically, don't think of the table as an object. Think of the table as a function that takes some input of raw data and produces some output that's the actual terminal characters that visually represents a table. Configuring the table is just modifying the function itself (colors, spacing, etc.). The function that then produces the table is pure: the same input produces the same output.

The model you're supplying is the input for the function. When you want to give a different input, you have to give a different input; this means resupplying all the model data at once, because that's the input. Whether this is changing a single value or otherwise doing a more expensive rebuild is outside the table's care. It just cares that you're giving it a new input, so it'll give a new output.

The optimizations of how it does this _quickly_ are hidden away from whatever's using this function: this is why you can't modify the data 'inside' the table, because there's conceptually not data 'inside' the table, and these optimizations only work _because_ we know the user isn't modifying those hidden caches.

I'm not sure if that made things more or less clear. :)

**Reply by timcolson** [reply]
Thanks for the additional info, @Evertras. I'm working on some experiments to solidify it in my brain. 👍
